### PR TITLE
[Discover-next] Add query editor extensions

### PR DESCRIPTION
**Pull request by joshuali925**
### Description

see #6894 

This PR picks #6894, #6895, #6933, #6972 to main. Additionally,
- separates extensions from query enhancements
- adds banner support
- partially revert #6972 as it's pending on the data source commit to main
- renames search bar extension to query editor extension

A query editor extension can display a UI component above the query editor and/or a banner above the language selector. The component has the ability to read and write discover search bar states to enhance the search experience for users. The configuration is part of UI Enhancements.

```ts
export interface QueryEditorExtensionDependencies {
  /**
   * Currently selected index patterns.
   */
  indexPatterns?: Array<IIndexPattern | string>;
  /**
   * Currently selected data source.
   */
  dataSource?: DataSource;
  /**
   * Currently selected query language.
   */
  language: string;
}

export interface QueryEditorExtensionConfig {
  /**
   * The id for the search bar extension.
   */
  id: string;
  /**
   * Lower order indicates higher position on UI.
   */
  order: number;
  /**
   * A function that determines if the search bar extension is enabled and should be rendered on UI.
   * @returns whether the extension is enabled.
   */
  isEnabled: (dependencies: QueryEditorExtensionDependencies) => Promise<boolean>;
  /**
   * A function that returns the search bar extension component. The component
   * will be displayed on top of the query editor in the search bar.
   * @param dependencies - The dependencies required for the extension.
   * @returns The component the search bar extension.
   */
  getComponent?: (dependencies: QueryEditorExtensionDependencies) => React.ReactElement | null;
  /**
   * A function that returns the search bar extension banner. The banner is a
   * component that will be displayed on top of the search bar.
   * @param dependencies - The dependencies required for the extension.
   * @returns The component the search bar extension.
   */
  getBanner?: (dependencies: QueryEditorExtensionDependencies) => React.ReactElement | null;
}

export interface UiEnhancements {
  query?: QueryEnhancement;
+ queryEditorExtension?: QueryEditorExtensionConfig;
}
```

### Issues Resolved

Developers can utilize search bar extensions to add additional features to the search bar, such as query assist.

Part of #6077

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/data/public/ui/query_editor/query_editor.tsx
Comment (kavilla) [inline src/plugins/data/public/ui/query_editor/query_editor.tsx:45]: sorry about that.

i might have ported this from your changes. did you prefer to use the header ref?\ I don't think we are using it for anything else.

File: src/plugins/discover/public/application/view_components/canvas/top_nav.tsx
Comment (kavilla) [inline src/plugins/discover/public/application/view_components/canvas/top_nav.tsx:92]: added an item here to follow up on this:

File: src/plugins/data/public/ui/search_bar/search_bar.tsx
Comment (kavilla) [inline src/plugins/data/public/ui/search_bar/search_bar.tsx:499]: quick question: do we know if this will have to be reset if toggling this feature off?

like if the select a datasource and toggle it off and that data source isn't available that should be fine right? or do we think in the future we will need to reset it like how i do here: https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/data/public/ui/settings/settings.ts#L52.

this 100% makes sense to me as you have it right now so I don't think we should change this because it's being passed in like index Patterns. but if we do think there might be implications on the toggle on and then off again, might be worth renaming the UI Settings service to something to else to be more descriptive and reset all that stuff in that service.

File: src/plugins/data/public/ui/search_bar/search_bar.tsx
Comment (kavilla) [inline src/plugins/data/public/ui/search_bar/search_bar.tsx:33]: nit: i think we have the isEqual `lodash` replaced. if you mind doing a quick check? but we had a goal to clean up lodash usage:

https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3660
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3899

but i would do this only if you have time. i'm guilty of adding more lodash as well.

File: src/plugins/data/public/ui/settings/settings.ts
Comment (kavilla) [inline src/plugins/data/public/ui/settings/settings.ts:168]: yeah @mengweieric called it [here](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/7001#discussion_r1638745574), once your stuff is merged I will convert this to an object as well.

File: src/plugins/data/public/index.ts
Comment (kavilla) [inline src/plugins/data/public/index.ts:530]: can be a fast follow: we should add it to the other query: https://github.com/opensearch-project/OpenSearch-Dashboards/blob/8b0af9de0215da61d23f4223bbce97a68d226e49/src/plugins/data/public/index.ts#L467

File: src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx
Comment (kavilla) [inline src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx:47]: nit: should we consider making this a behavior subject? then appending $ like `isEnabled$` might be easier to distinguish that i should subscribe to it.

File: src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx
Comment (kavilla) [inline src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx:54]: just noting here so don't change anything but i wonder if we should be consist with the return when related `undefined` or null.

https://github.com/opensearch-project/OpenSearch-Dashboards/pull/7001#discussion_r1637006688

any thoughts or preference? like i said, don't modify this and i didn't really have a good reason for undefined either so any opinion will be cool. if don't have a chance then that's cool.

File: src/plugins/data/public/ui/search_bar/search_bar.tsx
Comment (kavilla) [inline src/plugins/data/public/ui/search_bar/search_bar.tsx:36]: nit: quick question. was this auto sorted? should be fine but something to watch out incase some folks working on the discover-next stuff have their own linting pref as well to avoid bigger deltas in the changes.

if this was we could also consider adding the rule here: https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/.eslintrc.js

File: src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx
Comment (kavilla) [inline src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx:75]: is this intentionally set to `true` or should it be defaulted to false?

**Review by kavilla — APPROVED** [review]
awesome !!! looks great.

have a couple comments and nits, but nothing blocking and could just be fast follows after merging this.

**Comment by kavilla** [comment]
@joshuali925 you just have to install the bot: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/7034#issuecomment-2168834915 to your github then within the Changelog section of the [PR description:](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/7034#issue-2354183997), you can add the information in the [correct format](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/.github/pull_request_template.md?plain=1#L24) there and the changelog will automatically be generated with no manually requirement which we then use for the release notes generation. More info: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5519

Plugins should be able to unboard this too. s/o @BigSamu

File: src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx
Comment (kavilla) [inline src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx:73]: nit: do we want to consider making this an `QueryEditorExtension` and placing the component into a folder of it's on in the query_editor? or if the name is intentional.

one thing in my [fast follows ](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/6957) is to clean up shared components between the query string input and the query editor. so anything shared will their own folder but anything specific/only used for that component I was thinking would exist the folder.

But I can see the argument being made that this component is not unique to the query editor itself.

File: src/plugins/data/public/ui/query_editor/query_editor.tsx
Comment (joshuali925) [reply src/plugins/data/public/ui/query_editor/query_editor.tsx:45]: i just picked my previous PRs and added 8b0af9d. what did you port that can replace this?

File: src/plugins/discover/public/application/view_components/canvas/top_nav.tsx
Comment (joshuali925) [reply src/plugins/discover/public/application/view_components/canvas/top_nav.tsx:92]: thanks

File: src/plugins/data/public/ui/ui_service.ts
Comment (joshuali925) [reply src/plugins/data/public/ui/ui_service.ts:49]: thanks good point, i forgot about the id, will refactor

File: src/plugins/data/public/ui/search_bar/search_bar.tsx
Comment (joshuali925) [reply src/plugins/data/public/ui/search_bar/search_bar.tsx:499]: i'm not sure what the behavior would be, i wasn't able to test with data source too much. will need to revisit when we have the test environment

File: src/plugins/data/public/ui/search_bar/search_bar.tsx
Comment (joshuali925) [reply src/plugins/data/public/ui/search_bar/search_bar.tsx:33]: the PR didn't touch lodash usage, this came from organizing imports of the previous version of search_bar.tsx, which had them imported on separate lines
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/a4aa682226b9d22d2c6e7a41a11a44225b6d7db1/src/plugins/data/public/ui/search_bar/search_bar.tsx#L31-L36

it might be better to do this in another PR so that the change is more isolated

File: src/plugins/data/public/ui/settings/settings.ts
Comment (joshuali925) [reply src/plugins/data/public/ui/settings/settings.ts:168]: i was thinking the same thing, i tried to convert `ConstructorParameters<typeof Settings>` to a map on the type level but typescript isn't that flexible and can't manipulate parameter names

File: src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx
Comment (joshuali925) [reply src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx:54]: react functional component type definition uses null so i put null
```ts
    interface FunctionComponent<P = {}> {
        (props: PropsWithChildren<P>, context?: any): ReactElement<any, any> | null;
        propTypes?: WeakValidationMap<P> | undefined;
        contextTypes?: ValidationMap<any> | undefined;
        defaultProps?: Partial<P> | undefined;
        displayName?: string | undefined;
    }
```

File: src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx
Comment (joshuali925) [reply src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx:47]: i haven't used observables much, a quick search tells me to use `const isEnabled$ = (dependencies) => from(isEnabled(dependencies));` which i can't see what the benefit is. could you elaborate more a bit?

File: src/plugins/data/public/ui/search_bar/search_bar.tsx
Comment (joshuali925) [reply src/plugins/data/public/ui/search_bar/search_bar.tsx:36]: yes they were auto sorted. but this came from the typescript language server, i'm not sure if there's a corresponding eslint rule. i agree it would be better to enforce a consistent rule on import order

File: src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx
Comment (joshuali925) [reply src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx:75]: goot catch that probably came from testing, it should be false. will revert it

File: src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx
Comment (joshuali925) [reply src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx:73]: i was thinking that this can also be put on the old query bar. but since currently it is rendered in the query editor and i don't think we have plan to add it somewhere else, i think it's a good point. will do the refactor

File: src/plugins/data/public/ui/query_editor/query_editor.tsx
Comment (kavilla) [reply src/plugins/data/public/ui/query_editor/query_editor.tsx:45]: i copied over this entire file and then cleaned up stuff that wasn't available or from other people's stuff but i accidentally left this in from your stuff.

File: src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx
Comment (kavilla) [reply src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx:73]: yeah that's one of the things i was guessing too. makes sense. not a blocker tho if you get a second approval and can be a clean up.

File: src/plugins/data/public/ui/ui_service.ts
Comment (joshuali925) [reply src/plugins/data/public/ui/ui_service.ts]: why record instead of map tho?

File: src/plugins/data/public/ui/query_editor/query_editor.tsx
Comment (joshuali925) [reply src/plugins/data/public/ui/query_editor/query_editor.tsx:45]: oh i see, git didn't complain since i was picking the same modification. but the extension uses the header ref and banner ref to mount, current code is intended

the banner one was added after previous PRs, because UX requires some callout above language selector

**Review by kavilla — APPROVED** [review]

**Review by abbyhu2000 — APPROVED** [review]

File: src/plugins/data/public/ui/query_editor/query_editor_extensions/query_editor_extensions.tsx
Comment (AMoo-Miki) [inline src/plugins/data/public/ui/query_editor/query_editor_extensions/query_editor_extensions.tsx:23]:
```suggestion
    if ('[object Object]' !== Object.prototype.toString.call(configMap) || Object.keys(configMap).length === 0) return [];
```
This will need linting though.

File: src/plugins/data/public/ui/query_editor/query_editor_top_row.tsx
Comment (AMoo-Miki) [inline src/plugins/data/public/ui/query_editor/query_editor_top_row.tsx:262]: nit: perf
```suggestion
    if (!(
      queryEditorHeaderRef.current &&
      queryEditorBannerRef.current &&
      queryLanguage &&
      shouldRenderQueryEditorExtensions()
    ))
```

File: src/plugins/data/public/ui/query_editor/query_editor_top_row.tsx
Comment (AMoo-Miki) [inline src/plugins/data/public/ui/query_editor/query_editor_top_row.tsx:308]: Do we trust that `queryEditorExtensionMap` is going to be a plain object? I am not sure where it is coming from and hence asking: could it possibly be manipulated in the metadata store?

File: src/plugins/data/public/ui/ui_service.ts
Comment (AMoo-Miki) [inline src/plugins/data/public/ui/ui_service.ts:48]: Should this maybe be?
```suggestion
        if (enhancements.queryEditorExtension?.id) {
```

File: src/plugins/data/public/ui/query_editor/query_editor_extensions/query_editor_extensions.tsx
Comment (joshuali925) [reply src/plugins/data/public/ui/query_editor/query_editor_extensions/query_editor_extensions.tsx:23]: thanks good point on `.length === 0`, i missed that

is the object check necessary or conventional in OSD?

File: src/plugins/data/public/ui/query_editor/query_editor_top_row.tsx
Comment (joshuali925) [reply src/plugins/data/public/ui/query_editor/query_editor_top_row.tsx:308]: https://github.com/opensearch-project/OpenSearch-Dashboards/blob/7f0e39eb9809c95b98069cc971611edc2cbbc62b/src/plugins/data/public/ui/ui_service.ts#L31

https://github.com/opensearch-project/OpenSearch-Dashboards/blob/7f0e39eb9809c95b98069cc971611edc2cbbc62b/src/plugins/data/public/ui/ui_service.ts#L48-L51

what is the metadata store?

File: src/plugins/data/public/ui/ui_service.ts
Comment (joshuali925) [reply src/plugins/data/public/ui/ui_service.ts:48]: id is enforced on the type level, is type checking not enough and we prefer to also add runtime checking in osd?

**Comment by kavilla** [comment]
Since this is disabled as default and @AMoo-Miki not blocking. Feel free to add to this issue with fast follows: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/6957#issue-2339259590 or create a new issue than you can handle fast follows from here.

File: src/plugins/data/public/ui/query_editor/query_editor_extensions/query_editor_extension.tsx
Comment (virajsanghvi) [inline src/plugins/data/public/ui/query_editor/query_editor_extensions/query_editor_extension.tsx:100]: nit: use braces even for single line if statements

File: src/plugins/data/public/ui/query_editor/query_editor_extensions/query_editor_extensions.tsx
Comment (virajsanghvi) [inline src/plugins/data/public/ui/query_editor/query_editor_extensions/query_editor_extensions.tsx:19]: nit: is there a better name for this given its both a collection of extensions as well as specifying their destinations (and its easy to confuse components that are one character apart). perhaps its a manager as its responsibilities could include managing what gets rendered if we want more rules than rendering everything

File: src/plugins/data/public/ui/query_editor/query_editor_extensions/query_editor_extensions.tsx
Comment (virajsanghvi) [inline src/plugins/data/public/ui/query_editor/query_editor_extensions/query_editor_extensions.tsx:14]: q: is there a reason this has to be a map vs an array?

File: src/plugins/data/public/ui/query_editor/query_editor_extensions/query_editor_extension.tsx
Comment (joshuali925) [reply src/plugins/data/public/ui/query_editor/query_editor_extensions/query_editor_extension.tsx:100]: i think this would be a good linter rule. I personally prefer one line if-returns, but i agree using braces is a better practice

i'll create an issue to track this and enable eslint curly, without linter there might be missed ones or mistakes

File: src/plugins/data/public/ui/query_editor/query_editor_extensions/query_editor_extensions.tsx
Comment (joshuali925) [reply src/plugins/data/public/ui/query_editor/query_editor_extensions/query_editor_extensions.tsx:19]: maybe `QueryEditorExtensionManager` and `QueryEditorExtension`? but I'm not very sure if a manager is something that should be rendered on the page, current impl needs `QueryEditor` to render `QueryEditorExtensions`. The one character naming existed in other places, like `suggestion_component` and `suggestions_component`, although i do find it to be confusing

File: src/plugins/data/public/ui/query_editor/query_editor_extensions/query_editor_extensions.tsx
Comment (joshuali925) [reply src/plugins/data/public/ui/query_editor/query_editor_extensions/query_editor_extensions.tsx:14]: consistency with existing code, see https://github.com/opensearch-project/OpenSearch-Dashboards/pull/7034#discussion_r1641543079

**Pull request merged**
